### ISSUE TITLE
[proposition] Custom view filters

### ISSUE THREAD

**Issue by slavafomin** [balderdashy/sails#3130]
Hello!

I think we need a mechanism to allow developers to easily add custom view filters to the application.

The filters could be stored like this: `/api/view-filters/toUpper.js` and have the following structure:

```javascript
module.exports = function toUpper (value) {
  return value.toUpperCase();
}
```

From the Sails point of view, we will need to iterate the directory and add all functions found in it to the templating engine during project initialization. I think that `consolidate.js` should provide the API to add filters to all the different templating engines, however, I'm not sure if it does. I've created an issue here: https://github.com/tj/consolidate.js/issues/205 to request this feature. Maybe we will need to give a hand there too.

I think that view filters are extremely useful and needed feature of the server-side web framework and it will greatly help with promoting it. Right now, this area is highly confusing, especially for a new developers (like me, he-he).

**Comment by slavafomin** [comment]
The developers of Consolidate [are willing](https://github.com/tj/consolidate.js/issues/205#issuecomment-126832224) to accept PR's related to unified filter control interface.

I've tried to implement a custom experimental hook in my project in order to add filters to nunjucks via consolidate, however I can't make it work. Here's the code:

```javascript
module.exports = function viewFilters (sails) {
  return {
    initialize: function (done) {
      sails.on('hook:views:loaded', function () {

        // It seems like Sails puts the Consolidate instance here.
        var cons = sails.config.views.engine.fn;

        // Making sure property exists.
        cons.requires = cons.requires || {};

        // Initializing nunjucks if required.
        if (!cons.requires.nunjucks) {
          var nunjucks = require('nunjucks');
          cons.requires.nunjucks = nunjucks.configure();
        }

        // Getting instance of nunjucks from consolidate.
        var njInstance = cons.requires.nunjucks;

        // Adding a custom filter via nunjucks API.
        njInstance.addFilter('foo', function () {
          return 'FOO';
        });

        // Signalling Sails that we are done here.
        return done();

      });
    }
  }
};
```

I'm getting this error: `Template render error: Error: filter not found: foo`. I'm just starting to learn internals of Sails and Consolidate. What I'm I doing wrong, what could be the problem?

**Comment by slavafomin** [comment]
After more thorough investigation it looks like Sails [saves handle](https://github.com/balderdashy/sails/blob/v0.11.0/lib/hooks/views/configure.js#L63) not to entire Consolidate instance, but to a rendering function for specific engine.

So it's not providing any access to the overall Consolidate instance, and without it we can't customize the engine (can't obtain it properly).

**I think the first step would be to expose access to Consolidate in third-party hooks.**

**Comment by slavafomin** [comment]
I've been able to implement a very basic hook that already works for *nunjucks* and loads view filters from a directory. The gist is here: https://gist.github.com/slavafomin/912a23238d8614c1ef4b. I've also outlined the to-do there.

**Comment by sailsbot** [comment]
Thanks for posting, @slavafomin.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
